src/shared/script/shared-code/suffixUtils.ts

```diff
@@ -0,0 +1,305 @@
+  // number of trailing zeros after first digit
+  assertInteger(num)
+  if (Math.abs(num) < 10) {
+    return 0
+  }
+  let nZeros = 0
+  while (num % 10 === 0) {
+    num /= 10
+    nZeros++
+  }
+  return nZeros
+}
+
+export function isNumberGroupEven(n: number): boolean {
+  // big numbers (see: https://hu.wikipedia.org/wiki/T%C3%ADz_hatv%C3%A1nyai)
+  const nZeros = trailingZeros(n)
+  const nGroups = Math.floor(nZeros / 3)
+  return nGroups % 2 === 0
+}
+
+export function pitch(n: number): string {
+  // check if number is high pitch (e/i) or low pitch (a/o/u)
+  assertInteger(n)
+  const nZeros = trailingZeros(n)
+  if (nZeros <= 1) {
+    const highPitch = nZeros === 0 ? [1, 2, 4, 5, 7, 9] : [1, 4, 5, 7, 9]
+    const lastDigit = getLastNonZeroDigit(n)
+    return highPitch.indexOf(lastDigit) >= 0 ? 'high' : 'low'
+  } else if (nZeros >= 3 && nZeros <= 5) {
+    return 'high'
+  }
+  return 'low'
+}
+
+export function suffixVowel(n: number, vowelType: string): string {
+  const lowPitch = pitch(n) === 'low'
+  let vowel = ''
+  const type = vowelType.split('').sort().join('')
+  if (type === 'ae') {
+    // nak/nek
+    vowel = lowPitch ? 'a' : 'e'
+  } else if (type === 'áé') {
+    // át/ét
+    vowel = lowPitch ? 'á' : 'é'
+  } else if (type === 'eoö') {
+    // szor/szer/ször
+    const lastDigit = getLastDigit(n)
+    vowel = lastDigit === 5 || lastDigit === 2 ? 'ö' : lowPitch ? 'o' : 'e'
+  } else if (type === 'óő') {
+    // ból/ből
+    vowel = lowPitch ? 'ó' : 'ő'
+  } else {
+    throw new Error(`Invalid vowel type: "${vowelType}"`)
+  }
+  return vowel
+}
+
+export function dativSuffix(n: number): string {
+  // suffix: at/et/öt/t
+  const nZeros = trailingZeros(n)
+  if (nZeros < 6) {
+    const lastDigit = getLastDigit(n)
+    const exceptions = {
+      0: 't',
+      2: 't',
+      5: 'öt',
+      6: 'ot',
+    }
+    // handle single digit exceptions
+    if (nZeros === 0 && lastDigit in exceptions) {
+      return exceptions[lastDigit]
+    } else {
+      return pitch(n) === 'high' ? 'et' : 'at'
+    }
+  } else if (nZeros === 9) {
+    return 'ot'
+  } else {
+    return 't'
+  }
+}
+
+export function placeValueSuffix(n: number, sfx: string): string {
+  // suffix: as/es/asoknak/eseket etc.
+  if (!/^[aáeoö]?s/g.test(sfx)) {
+    throw new Error(`Invalid suffix: "${sfx}"`)
+  }
+  const lastDigit = getLastDigit(n)
+  const lowPitch = pitch(n) === 'low'
+  const nZeros = trailingZeros(n)
+
+  // special cases
+  let replacement = ''
+  if (nZeros < 6) {
+    if (n === 0) {
+      replacement = ''
+    } else if (lastDigit === 5) {
+      replacement = 'ö'
+    } else if (lastDigit === 6) {
+      replacement = 'o'
+    } else {
+      replacement = lowPitch ? 'a' : 'e'
+    }
+  } else {
+    replacement = isNumberGroupEven(n) ? '' : 'o'
+  }
+  sfx = sfx.replace(/^[aáeoö]?(?=s)/g, replacement) // -s/as/es/os/ös
+  sfx = sfx.replace(/(?<=s)[eoö](?=k)/g, suffixVowel(n, 'oeö')) // -sok/sek/sök
```

Comment:
I do not understand why do you need "lookbehind assertions" here
```suggestion
  sfx = sfx.replace(/^(s)[eoö](k)$/g, `$1${suffixVowel(n, 'oeö')}$2`) // -sok/sek/sök
```

Reply:
done.